coderd/insights.go

```diff
@@ -531,20 +531,37 @@ func parseInsightsStartAndEndTime(ctx context.Context, rw http.ResponseWriter, s
 	return startTime, endTime, true
 }
 
-func verifyInsightsInterval(ctx context.Context, rw http.ResponseWriter, intervalString string) (codersdk.InsightsReportInterval, bool) {
+func parseInsightsInterval(ctx context.Context, rw http.ResponseWriter, intervalString string, startTime, endTime time.Time) (codersdk.InsightsReportInterval, bool) {
 	switch v := codersdk.InsightsReportInterval(intervalString); v {
 	case codersdk.InsightsReportIntervalDay, "":
 		return v, true
+	case codersdk.InsightsReportIntervalWeek:
+		if !lastReportIntervalHasAtLeastSixDays(startTime, endTime) {
+			httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+				Message: "Query parameter has invalid value.",
+				Detail:  "Last report interval should have at least 6 days.",
+			})
+			return "", false
+		}
+		return v, true
 	default:
 		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
 			Message: "Query parameter has invalid value.",
 			Validations: []codersdk.ValidationError{
 				{
 					Field:  "interval",
-					Detail: fmt.Sprintf("must be one of %v", []codersdk.InsightsReportInterval{codersdk.InsightsReportIntervalDay}),
+					Detail: fmt.Sprintf("must be one of %v", []codersdk.InsightsReportInterval{codersdk.InsightsReportIntervalDay, codersdk.InsightsReportIntervalWeek}),
 				},
 			},
 		})
 		return "", false
 	}
 }
+
+func lastReportIntervalHasAtLeastSixDays(startTime, endTime time.Time) bool {
+	lastReportIntervalDays := endTime.Sub(startTime) % (7 * 24 * time.Hour)
+	if lastReportIntervalDays == 0 {
+		return true // this is a perfectly full week!
+	}
+	return lastReportIntervalDays >= 6*24*time.Hour
+}
```

Comment:
A test case for such a scenario may be good.

Reply:
I can see an example where:
1. `startTime` is before the DST transition
2. `endTime` is after the transition but exactly 6 "days" apart according to the standard 24-hour day.

The actual time elapsed between `startTime` and `endTime` in this case will be less than 6 * 24 hours due to the one-hour shift forward.

This "6 days" period is an exception for the current (and last) interval, and could be potentially an issue for an hour. I'm not sure what is the best way to fix this. Should I add an extra condition:

`return lastReportIntervalDays >= 6*24*time.Hour || startTime.AddDate(0, 0, 6).Equal(endTime)`

Reply:
Why not instead just check `startTime.AddDate(0, 0, 6).Equal(endTime)` instead of checking the hours? That should be 'correct' for all cases, unless I'm missing something?

Reply:
I added a test to verify function `lastReportIntervalHasAtLeastSixDays`.